### PR TITLE
Refactor partition groups and add tests

### PR DESCRIPTION
**Pull request by jayshrivastava**
This change adds a new abstraction called a `Grouper` in `distribution_strategy.rs`. This is responsible for grouping partitions into partition groups. The new `distribution_strategy` module can be used to store other splitters and, in the future, distribution strategies like round robin etc.

A "partition group" is now a range [starting_partition, ending_partition) instead of a Vec, so we improve space efficiency. In the proto layer, the `DDTask` proto still uses a Vec, so we simply expand the range into a Vec when creating them in `planning.rs::assign_to_workers`.

This also fixes a bug in the isolator where it may not have reported the right number of partitions.

Testing
- unit tests in isolator.rs
- unit tests for `build_replacement` in `planning.rs`
- unit tests for the grouper in `distribution_strategy.rs`

### CODE REVIEW

File: src/planning.rs [hunk 1]
```diff
@@ -585,22 +596,23 @@ fn assign_to_workers(
...
                 }
             }
 
+            // TODO: It would be more efficient to avoid materializing the partition group below.
             let task_data = DDTask {
                 query_id: query_id.to_string(),
                 stage_id: stage.stage_id,
                 plan_bytes,
-                partition_group: partition_group.to_vec(),
+                partition_group: (partition_group.start() as u64..partition_group.end() as u64).collect(),
```
Comment (jayshrivastava) [inline src/planning.rs]: Please let me know if I can change the `DDTask` proto to use a range as well.

File: src/planning.rs [hunk 2]
```diff
@@ -669,38 +681,31 @@ fn get_stage_addrs_from_tasks(target_stage_ids: &[u64], stages: &[DDTask]) -> Re
...
-            .collect(),
-        None => vec![(0..partition_count).collect()],
-    };
+    let partition_groups = splitter.split(partition_count);
 
     if isolate && partition_groups.len() > 1 {
         let new_child = Arc::new(PartitionIsolatorExec::new(
             child.clone(),
-            partitions_per_worker.unwrap(), // we know it is a Some, here.
+            partition_count,
```
Comment (jayshrivastava) [inline src/planning.rs]: I think this was a bug before. The isolator seems to want the total number of partitions here.

File: src/codec.rs [hunk 3]
```diff
@@ -305,7 +305,7 @@ mod test {
         let schema = create_test_schema();
         let part = Partitioning::UnknownPartitioning(2);
         let reader_exec = Arc::new(DDStageReaderExec::try_new(part, schema, 1).unwrap());
-        let exec = Arc::new(PartitionIsolatorExec::new(reader_exec, 4));
+        let exec = Arc::new(PartitionIsolatorExec::new(reader_exec, 2));
```
Comment (NGA-TRAN) [inline src/codec.rs]: We probably don’t need to change the value from 4 to 2 in this test, since having just 2 input partitions is valid. Given our setup—where each worker can handle up to 4 partitions—those 2 inputs will be sent to and processed by a single worker.

On a related note, this raises the idea of adding tests that vary the number of input and output partitions. It might be worth establishing a best practice of opening a ticket whenever we identify follow-up work. Let me know if you'd like me to create one for this, or if you'd prefer to take the lead.

File: src/isolator.rs [hunk 4]
```diff
@@ -73,9 +73,9 @@ impl ExecutionPlan for PartitionIsolatorExec {
     }
 
     fn with_new_children(
-        self: std::sync::Arc<Self>,
-        children: Vec<std::sync::Arc<dyn ExecutionPlan>>,
-    ) -> Result<std::sync::Arc<dyn ExecutionPlan>> {
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
```
Comment (NGA-TRAN) [inline src/isolator.rs:84]: ❤️ Thanks for simplifying the code

File: src/distribution_strategy.rs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        self.start
+    }
+
+    // end is the exclusive end partition in the range
+    pub fn end(&self) -> usize {
+        self.end
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
```
Comment (NGA-TRAN) [inline src/distribution_strategy.rs:38]: Can you also add comment explaining the purpose of this struct? It is clear from this PR what it is for but, in the future, if readers only come to this struct, they won't have context and the comment will help

File: src/distribution_strategy.rs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    fn test_partition_grouper_uneven() {
+        let grouper = PartitionGrouper::new(2);
+        let groups = grouper.split(5);
+
+        let expected = vec![
+            PartitionGroup { start: 0, end: 2 },
+            PartitionGroup { start: 2, end: 4 },
+            PartitionGroup { start: 4, end: 5 },
+        ];
+
```
Comment (NGA-TRAN) [inline src/distribution_strategy.rs:96]: Can you add these tests:
1. `PartitionGrouper::new(2)` and `grouper.split(1);`. Idea: number of input partitions < number of partitions a work is set up to handle
2. `PartitionGrouper::new(2)` and `grouper.split(2);`. Idea: they are the same
3. `PartitionGrouper::new(2)` and `grouper.split(0);`. Idea: no input partitions or empty partition. Check the code to verify if empty partition is accepted. IIRC, this is a valid case and we need to see how we handle this in distributed plan

File: src/planning.rs [hunk 7]
```diff
@@ -240,10 +241,10 @@ pub async fn physical_planning(
 }
 
 /// Returns distributed plan and execution stages for both query execution and EXPLAIN display
-pub async fn distributed_physical_planning(
+pub async fn distributed_physical_planning<S: Grouper + Copy>(
     physical_plan: Arc<dyn ExecutionPlan>,
+    splitter: S,
```
Comment (NGA-TRAN) [inline src/planning.rs]: I suggest to make the name more explicit. Maybe `partition_splitter`? or even more meaningful. Can you also add comment to explain that this splitter is for?

File: src/query_planner.rs [hunk 8]
```diff
@@ -148,18 +149,18 @@ impl QueryPlanner {
 
         // construct the distributed physical plan and stages
         let (distributed_plan, distributed_stages) =
-            distributed_physical_planning(physical_plan.clone(), 8192, Some(2)).await?;
+            distributed_physical_planning(physical_plan.clone(), PartitionGrouper::new(2), 8192)
```
Comment (NGA-TRAN) [inline src/query_planner.rs:152]: **FYI:** We have this ticket to remind us to work on global config https://github.com/datafusion-contrib/datafusion-distributed/issues/23. These number 2 and 8192 should be a part of the config work which can be either in the same ticket or a separate and new ticket. If you think these should be in need a separate ticket, please feel free to create one or let me know, I am happy to do it top remind us to do so

File: src/isolator.rs [hunk 9]
```diff
@@ -143,3 +143,73 @@ impl ExecutionPlan for PartitionIsolatorExec {
...
+        }
+
+        let task_context = ctx.task_ctx();
+
+        // Success case: execute valid partition
+        let result = isolator.execute(0, task_context.clone());
+        assert!(result.is_ok());
+
+        // Error case: try to execute partition beyond partition_count
+        let result = isolator.execute(4, task_context.clone());
```
Comment (NGA-TRAN) [inline src/isolator.rs:207]: 👍

File: src/isolator.rs [hunk 10]
```diff
@@ -143,3 +143,73 @@ impl ExecutionPlan for PartitionIsolatorExec {
...
+        let empty_ctx = SessionContext::new();
+        let empty_task_context = empty_ctx.task_ctx();
+
+        let result = isolator.execute(0, empty_task_context);
+        assert!(result.is_err());
+        if let Err(e) = result {
+            assert!(e
+                .to_string()
+                .contains("PartitionGroup not set in session config"));
+        }
```
Comment (NGA-TRAN) [inline src/isolator.rs:269]: Excellent test. Can you also add `let result = isolator.execute(1, empty_task_context);` right here?

File: src/isolator.rs [hunk 11]
```diff
@@ -143,3 +143,73 @@ impl ExecutionPlan for PartitionIsolatorExec {
...
+    }
+}
```
Comment (NGA-TRAN) [inline src/isolator.rs:271]: Can you add a few more tests with different value of `partition_count` and `partition_group`. Make sure enough combination with different values of them. You have tested their count `equal`. Adding `less` and `greater` will help capture bugs. Think about all kinds of edge cases

File: src/planning.rs [hunk 12]
```diff
@@ -309,7 +310,7 @@ pub async fn distributed_physical_planning(
             // left side of the join and is not suitable to be executed in a
             // partitioned manner.
             let mut replacement = plan.clone();
-            let partition_count = plan.output_partitioning().partition_count() as u64;
+            let partition_count = plan.output_partitioning().partition_count();
```
Comment (NGA-TRAN) [inline src/planning.rs:313]:
```suggestion
            let partition_group_size = 1; // NestedLoopJoinExec must be on a stage by itself
            let partition_count = plan.output_partitioning().partition_count();
```

File: src/planning.rs [hunk 13]
```diff
@@ -318,15 +319,15 @@ pub async fn distributed_physical_planning(
                 batch_size,
             )) as Arc<dyn ExecutionPlan>;
 
-            partition_groups = vec![(0..partition_count).collect()];
+            partition_groups = PartitionGrouper::new(1).split(partition_count);
```
Comment (NGA-TRAN) [inline src/planning.rs]:
```suggestion
            partition_groups = PartitionGrouper::new(partition_group_size).split(partition_count);
```

File: src/planning.rs [hunk 14]
```diff
@@ -318,15 +319,15 @@ pub async fn distributed_physical_planning(
...
             full_partitions = true;
             Ok(Transformed::yes(replacement))
         } else {
             trace!("not special case partition_groups: {:?}", partition_groups);
 
-            let partition_count = plan.output_partitioning().partition_count() as u64;
+            let partition_count = plan.output_partitioning().partition_count();
             // set this back to default
-            partition_groups = vec![(0..partition_count).collect()];
+            partition_groups = PartitionGrouper::new(1).split(partition_count);
```
Comment (NGA-TRAN) [inline src/planning.rs]: Ditto: assign a value as above next to the comment to say why it is 1

File: src/planning.rs [hunk 15]
```diff
@@ -585,22 +586,24 @@ fn assign_to_workers(
...
                         .entry(stage.stage_id)
                         .or_default()
-                        .entry(*part)
+                        .entry(part as u64)
                         .or_default()
                         .push(host.clone());
                 }
             }
 
+            // TODO: It would be more efficient to avoid materializing the partition group below.
```
Comment (NGA-TRAN) [inline src/planning.rs]: 👍

File: src/planning.rs [hunk 16]
```diff
@@ -669,39 +672,29 @@ fn get_stage_addrs_from_tasks(target_stage_ids: &[u64], stages: &[DDTask]) -> Re
 }
 
 #[allow(clippy::type_complexity)]
-fn build_replacement(
+fn build_replacement<S: Grouper>(
     plan: Arc<dyn ExecutionPlan>,
-    partitions_per_worker: Option<usize>,
+    splitter: S,
```
Comment (NGA-TRAN) [inline src/planning.rs]: Same as above. Make the name more explicit

File: src/planning.rs [hunk 17]
```diff
@@ -713,3 +706,75 @@ fn build_replacement(
...
+        let repartition = coalesce.children()[0];
+        let isolator = repartition.children()[0]
+            .as_any()
+            .downcast_ref::<PartitionIsolatorExec>()
+            .unwrap();
+        assert_eq!(
+            isolator.partition_count,
+            plan.properties().partitioning.partition_count()
+        )
+    }
```
Comment (NGA-TRAN) [inline src/planning.rs:779]: I am very interested in a lot more tests here.  Either you can add int his PR or in a different/follow-on PR. Here is some ideas:

1. Add great combination of inputs to the function `build_replacement` so we know they work well. Try edge cases
2. Make some mock data test if needed so instead of 1000 max row count or 8192 batch size, make them a lot smaller to control number of partitions out we want to test

File: src/query_planner.rs (hunk 8 above)
Comment (jayshrivastava) [reply src/query_planner.rs:152]: Created a new sub-issue on #23
